### ISSUE TITLE
[MCOMPILER-224] Maven compiler plugin does not properly consume plexus compiler output

### ISSUE THREAD

**Issue by olamy** [olamy/maven-compiler-plugin#32]
**[David M. Lloyd](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dmlloyd)** opened **[MCOMPILER-224](https://issues.apache.org/jira/browse/MCOMPILER-224?redirect=false)** and commented

Maven compiler plugin does not properly read the output from the plexus compiler.  All CompilerMessages are logged at warning or error level, even if they are not warnings or errors.

---

**Affects:** 3.0, 3.1

**Attachments:**
- [MCOMPILER-224.patch](https://issues.apache.org/jira/secure/attachment/12714910/MCOMPILER-224.patch) (_2.85 kB_)
- [MCOMPILER-224.zip](https://issues.apache.org/jira/secure/attachment/12714877/MCOMPILER-224.zip) (_7.16 kB_)

**Issue Links:**
- [MCOMPILER-66](https://issues.apache.org/jira/browse/MCOMPILER-66) Compiler swallows messages from annotation processors
 (_**"supercedes"**_)

6 votes, 5 watchers

**Comment by olamy** [comment]
**[David M. Lloyd](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dmlloyd)** commented

Uploaded patch.

**Comment by olamy** [comment]
**[David M. Lloyd](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dmlloyd)** commented

I think MCOMPILER-66 is probably out of date, as processor output isn't necessarily being swallowed right now, except maybe in error cases (however it is being logged wrong).

**Comment by olamy** [comment]
**[David M. Lloyd](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dmlloyd)** commented

FYI [John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey) is currently reviewing this.

**Comment by olamy** [comment]
**[David M. Lloyd](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dmlloyd)** commented

Looking for any update here.

**Comment by olamy** [comment]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** commented

proposed integration test

**Comment by olamy** [comment]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** commented

I've checked the patch against the integration test I just attached, and using a build of plexus-compiler snapshots from last Friday...it doesn't work.

I've debugged this down to the DiagnosticCollector class from the JDK, and it doesn't seem to collect those NOTE | OTHER | WARNING messages...only MANDATORY_WARNING and ERROR. I feel like I'm missing something here, since I had heard that a code fix was applied to plexus-compilers for this.

**Comment by olamy** [comment]
**[David M. Lloyd](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dmlloyd)** commented

What version of plexus-compiler are you using?

**Comment by olamy** [comment]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** commented

\<plexusCompilerVersion>2.4-SNAPSHOT\</plexusCompilerVersion>

and the build of that is from 03c0a61ed3f3fcdd73c41d34593dac8d03a6b4be on the master branch of:

https://github.com/sonatype/plexus-compiler.git

**Comment by olamy** [comment]
**[David M. Lloyd](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dmlloyd)** commented

I ask because it doesn't sound like it's using the embedded javac.  The interface to the external compiler doesn't support gathering other levels.

**Comment by olamy** [comment]
**[David M. Lloyd](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dmlloyd)** commented

If you add this line to the pom:

```xml
<plugin>
  <groupId>org.apache.maven.plugins</groupId>
  <artifactId>maven-compiler-plugin</artifactId>
  <version>3.2-SNAPSHOT</version>
    <configuration>
        <showWarnings>true</showWarnings> <!-- this line -->
    </configuration>
</plugin>
```

Then it works.  I think it has to do with the way plexus-compiler forwards output... I can look into that separately.

**Comment by olamy** [comment]
**[David M. Lloyd](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dmlloyd)** commented

Actually it seems to just be a basic behavior of the compiler itself.  So you have to have that option to see warning (and info) output.

**Comment by olamy** [comment]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** commented

Added IT and verified fix.

**Comment by olamy** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

the IT is fine with JDK 7 & 8, but fails with JDK 6 because the mesage isn't exactly the same:

jdk6:

```
[INFO] Compiling 1 source file to /home/herve/projets/maven/trunks/plugins/maven-compiler-plugin/target/it/MCOMPILER-224/annotation-user/target/classes
Testing message for: WARNING
Testing message for: MANDATORY_WARNING
Testing message for: NOTE
Testing message for: OTHER
[DEBUG] incrementalBuildHelper#afterRebuildExecution
[WARNING] warning: WARNING Test message.
[WARNING] warning: MANDATORY_WARNING Test message.
[INFO] Note: NOTE Test message.
[INFO] Note: OTHER Test message.
```

jdk7 & 8:

```
[INFO] Compiling 1 source file to /home/herve/projets/maven/trunks/plugins/maven-compiler-plugin/target/it/MCOMPILER-224/annotation-user/target/classes
Testing message for: WARNING
Testing message for: MANDATORY_WARNING
Testing message for: NOTE
Testing message for: OTHER
[DEBUG] incrementalBuildHelper#afterRebuildExecution
[WARNING] bootstrap class path not set in conjunction with -source 1.6
[WARNING] WARNING Test message.
[WARNING] MANDATORY_WARNING Test message.
[INFO] NOTE Test message.
[INFO] OTHER Test message.
```

can you check if the IT should be relaxed or there is some fix to remove the  extra "warning:" and "Note:" prefix, please?

**Comment by olamy** [comment]
**[David M. Lloyd](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dmlloyd)** commented

I think it'd just have to be a string transformation, special-cased for that JDK...

**Comment by olamy** [comment]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** commented

Sorry for the delay, I'm picking this back up. If the fix is already in, I'll simply verify that fact and close.

**Comment by olamy** [comment]
**[David M. Lloyd](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dmlloyd)** commented

I haven't provided any patch to fix the JDK 6 test problem, but I can, if you would like.

**Comment by olamy** [comment]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** commented

**sigh** apparently email sync doesn't work, so rewriting my comment here...

I thought I peeked at the source some time ago, and a little elf (Herve?) had been in there fixing up the JDK 6 handling... ;-D

I'll verify and adjust as needed.

**Comment by olamy** [comment]
**[David M. Lloyd](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dmlloyd)** commented

OK cool, just ping me if you need anything more.

**Comment by olamy** [comment]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** commented

I've added a cleanup step to the construction of CompilerMessage in plexus-compiler-api to address the JDK 6 message prefix problem causing the IT for this issue to fail.

The PR for plexus-compiler is: https://github.com/sonatype/plexus-compiler/pull/24

**Comment by olamy** [comment]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** commented

fixed plexus-compiler-api (PR#24) in order to clean up message text in CompilerMessage instances, and adjusted integration test to make sure everything is working.

**Comment by olamy** [comment]
**[David M. Lloyd](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dmlloyd)** commented

Thanks John!
